### PR TITLE
Persist element definitions

### PR DESCRIPTION
**Pull request by k88hudson**
This is working for images and text so far.

### CODE REVIEW

File: www_src/blocks/generator.js [hunk 1]
```diff
@@ -11,18 +12,19 @@ Generator.IMAGE = "image";
 
 Generator.generateBlock = function(options) {
```
Comment (k88hudson) [inline www_src/blocks/generator.js]: Personally I would rather get rid of this function altogether and just export the block components

Comment (k88hudson) [inline www_src/blocks/generator.js]: Actually eyah i think i'm gonna do that

File: package.json [hunk 2]
```diff
@@ -62,7 +62,7 @@
     "react": "0.13.1",
     "react-hammerjs": "0.2.2",
     "react-imageloader": "1.2.0",
-    "url-pattern": "^0.8.0",
+    "url-pattern": "git://github.com/k88hudson/url-pattern.git#hyphenunderscore",
```
Comment (Pomax) [inline package.json]: what does this change and is there already something that does this so we don't have to point to personal git repos?

File: www_src/blocks/generator.js [hunk 3]
```diff
@@ -2,6 +2,7 @@ var React = require('react');
 var Text = require('./text.jsx');
 var Link = require('./link.jsx');
 var Image = require('./image.jsx');
+var uuid = require('../lib/uuid');
```
Comment (Pomax) [inline www_src/blocks/generator.js]: why not https://www.npmjs.com/package/node-uuid?

File: www_src/blocks/generator.js [hunk 4]
```diff
@@ -11,18 +12,19 @@ Generator.IMAGE = "image";
...
-    return <Text value={options.value} />;
+    return <Text {...options} />;
   }
   if (options.type === Generator.LINK) {
-   return <Link href={options.href} label={options.label} active={options.active} />;
+   return <Link {...options} />;
   }
   if (options.type === Generator.IMAGE) {
-    return <Image src={options.src} alt={options.alt} />;
+    return <Image {...options} />;
```
Comment (Pomax) [inline www_src/blocks/generator.js]: Does ES6 +JSX give us some way to compact this to something like `<${type} {...options}/>`? =/

File: www_src/blocks/image.jsx [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    defaults: {
+      src: '',
+      alt: '',
+      opacity: 1,
+      borderWidth: 0,
+      borderColor: 'transparent'
+    }
+  },
+  getDefaultProps: function () {
+    return this.defaults;
```
Comment (Pomax) [inline www_src/blocks/image.jsx]: since it's statics, `return Image.defaults`?

File: www_src/blocks/image.jsx [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+  },
+  render: function() {
+    var style = {
+      opacity: this.props.opacity,
+      borderStyle: 'solid',
+      borderWidth: this.props.borderWidth,
+      borderColor: this.props.borderColor
+    };
+    return <img style={style} src={this.props.src} alt={this.props.alt} />
```
Comment (Pomax) [inline www_src/blocks/image.jsx]: do we want to use the slightly more complex LoaderImage that @thisandagain created?

File: www_src/blocks/text.jsx [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+      fontWeight: 'normal',
+      fontStyle: 'normal',
+      textDecoration: 'none',
+      textAlign: 'center',
+      innerHTML: 'Hello world'
+    }
+  },
+
+  getDefaultProps: function () {
+    return this.defaults;
```
Comment (Pomax) [inline www_src/blocks/text.jsx]: see above

File: www_src/blocks/text.jsx [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  },
+
+  render: function() {
+    var style = {};
+    ['fontSize', 'fontFamily', 'color', 'fontWeight', 'fontStyle', 'textDecoration', 'textAlign'].forEach(prop => style[prop] = this.props[prop]);
```
Comment (Pomax) [inline www_src/blocks/text.jsx]: `Object.keys(Text.defaults)`.forEach? To keep things DRY?

File: www_src/blocks/generator.js (hunk 3 above)
Comment (k88hudson) [inline www_src/blocks/generator.js]: we already have this in the repo

File: package.json (hunk 2 above)
Comment (k88hudson) [inline package.json]: I had to land some changes, hopefully they'll be merged by next week

File: www_src/blocks/image.jsx (hunk 6 above)
Comment (k88hudson) [inline www_src/blocks/image.jsx]: not in this patch preferably

File: www_src/blocks/image.jsx (hunk 5 above)
Comment (k88hudson) [inline www_src/blocks/image.jsx]: `Image` is undefined at this point, also it appears statics get attached to instances as well, so what's the downside?

File: www_src/blocks/text.jsx (hunk 8 above)
Comment (k88hudson) [inline www_src/blocks/text.jsx]: the problem is `innerHTML` is not included here. once we refactor to have style and attributes in separate objects this should be a lot cleaner.

File: www_src/components/range/range.jsx [hunk 9]
```diff
@@ -18,10 +19,11 @@ var Range = React.createClass({
   render: function () {
     var linkState = this.props.linkState || this.linkState;
     var valueLink = linkState(this.props.id);
+
     return (
       <div className="range">
-        <input min={this.props.min} max={this.props.max} type="range" valueLink={valueLink}/>
-        <div className={'range-summary' + (parseInt(valueLink.value, 0) === this.props.min ? ' min' : '')}>{valueLink.value}{this.props.unit}</div>
+        <input min={this.props.min} max={this.props.max} step={this.props.step} type="range" valueLink={valueLink}/>
+        <div className={'range-summary' + (parseFloat(valueLink.value, 10) === this.props.min ? ' min' : '')}>{valueLink.value}{this.props.unit}</div>
```
Comment (k88hudson) [inline www_src/components/range/range.jsx]: good call

File: package.json (hunk 2 above)
Comment (k88hudson) [inline package.json]: if you can think of a better alternative i'm all for it

File: www_src/blocks/image.jsx (hunk 6 above)
Comment (thisandagain) [inline www_src/blocks/image.jsx]: We can add it in a follow-up. The loading state still hasn't been spec'd.

**Comment by thisandagain** [comment]
This is working fairly well for me as long as I use the "soft" back button which triggers a new `onLoad` event for the parent activity. If you use the Android "hard" back button the prior view is shown without the changes being reflected. Related to #1695

File: package.json (hunk 2 above)
Comment (Pomax) [inline package.json]: just worried we never remember to switch back. Do we have a tracking issue for this?

File: www_src/blocks/image.jsx (hunk 5 above)
Comment (Pomax) [inline www_src/blocks/image.jsx]: when this function gets called, Image will be defined.

**Pull request merged**